core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -751,6 +751,14 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
    * Visible for testing
    */
   private[history] def doMergeApplicationListing(
+      reader: EventLogFileReader,
+      scanTime: Long,
+      enableOptimizations: Boolean,
+      lastEvaluatedForCompaction: Option[Long]): Unit = mergeApplicationListing(
+    reader, scanTime, enableOptimizations, lastEvaluatedForCompaction)
+
+  @scala.annotation.tailrec
+  private def mergeApplicationListing(
```

Comment:
Otherwise:

```
[error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:754:24: could not optimize @tailrec annotated method doMergeApplicationListing: it is neither private nor final so can be overridden
[error]   private[history] def doMergeApplicationListing(
[error]                        ^
[error] one error found
[error] (core / Compile / compileIncremental) Compilation failed
[error] Total time: 204 s (03:24), completed Dec 30, 2021 2:54:38 AM
```